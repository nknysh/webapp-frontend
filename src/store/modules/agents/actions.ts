import { ICompany, ITravelAgent } from 'services/BackendApi';
export const GET_TRAVEL_AGENTS_REQUEST = 'agentsModule/GET_TRAVEL_AGENTS_REQUEST';
export const GET_TRAVEL_AGENTS_SUCCESS = 'agentsModule/GET_TRAVEL_AGENTS_SUCCESS';
export const GET_TRAVEL_AGENTS_FAILURE = 'agentsModule/GET_TRAVEL_AGENTS_FAILURE';
export const SELECTED_TA_CHANGE = 'agentsModule/SELECTED_TA_CHANGE';
export const SHOW_TA_DROPDOWN = 'agentsModule/SHOW_TA_DROPDOWN';
export const SEARCH_TA_BY_NAME_CHANGE = 'agentsModule/SEARCH_TA_BY_NAME_CHANGE';
export const GET_COMPANIES_REQUEST = 'agentsModule/GET_COMPANIES_REQUEST';
export const GET_COMPANIES_SUCCESS = 'agentsModule/GET_COMPANIES_SUCCESS';
export const GET_COMPANIES_FAILURE = 'agentsModule/GET_COMPANIES_FAILURE';
export const SELECTED_COMPANY_CHANGE = 'agentsModule/SELECTED_COMPANY_CHANGE';
export const SHOW_COMPANY_DROPDOWN = 'agentsModule/SHOW_COMPANY_DROPDOWN';
export const SEARCH_COMPANY_BY_NAME = 'agentsModule/SEARCH_COMPANY_BY_NAME';

export type GetTravelAgentRequestAction = ReturnType<typeof getTravelAgentsRequestAction>;
export const getTravelAgentsRequestAction = (forceLoad?: boolean) => ({
  type: GET_TRAVEL_AGENTS_REQUEST as typeof GET_TRAVEL_AGENTS_REQUEST,
  forceLoad,
});

export type GetTravelAgentSuccessAction = ReturnType<typeof getTravelAgentsSuccessAction>;
export const getTravelAgentsSuccessAction = (agents: ITravelAgent[]) => ({
  type: GET_TRAVEL_AGENTS_SUCCESS as typeof GET_TRAVEL_AGENTS_SUCCESS,
  agents,
});

export type GetTravelAgentFailureAction = ReturnType<typeof getTravelAgentsFailureAction>;
export const getTravelAgentsFailureAction = (error: string) => ({
  type: GET_TRAVEL_AGENTS_FAILURE as typeof GET_TRAVEL_AGENTS_FAILURE,
  error,
});

export type SelectedTaChangeAction = ReturnType<typeof selectedTaChangeAction>;
export const selectedTaChangeAction = (value: ITravelAgent | null) => ({
  type: SELECTED_TA_CHANGE as typeof SELECTED_TA_CHANGE,
  value,
});

export type ShowTaDropdownAction = ReturnType<typeof showTaDropdownAction>;
export const showTaDropdownAction = (value: boolean) => ({
  type: SHOW_TA_DROPDOWN as typeof SHOW_TA_DROPDOWN,
  value,
});

export type SearchTaByNameChangeAction = ReturnType<typeof searchTaByNameChangeAction>;
export const searchTaByNameChangeAction = (value: string) => ({
  type: SEARCH_TA_BY_NAME_CHANGE as typeof SEARCH_TA_BY_NAME_CHANGE,
  value,
});


export type GetCompaniesRequestAction = ReturnType<typeof getCompaniesRequestAction>;
export const getCompaniesRequestAction = () => ({
  type: GET_COMPANIES_REQUEST as typeof GET_COMPANIES_REQUEST,
});

export type GetCompaniesSuccessAction = ReturnType<typeof getCompaniesSuccessAction>;
export const getCompaniesSuccessAction = (companies: ICompany[]) => ({
  type: GET_COMPANIES_SUCCESS as typeof GET_COMPANIES_SUCCESS,
  companies,
});

export type GetCompaniesFailureAction = ReturnType<typeof getCompaniesFailureAction>;
export const getCompaniesFailureAction = (error: string) => ({
  type: GET_COMPANIES_FAILURE as typeof GET_COMPANIES_FAILURE,
  error,
});

export type SelectedCompanyChangeAction = ReturnType<typeof selectedCompanyChangeAction>;
export const selectedCompanyChangeAction = (value: ICompany) => ({
  type: SELECTED_COMPANY_CHANGE as typeof SELECTED_COMPANY_CHANGE,
  value,
});

export type ShowCompanyDropdownAction = ReturnType<typeof showCompanyDropdownAction>;
export const showCompanyDropdownAction = (value: boolean) => ({
  type: SHOW_COMPANY_DROPDOWN as typeof SHOW_COMPANY_DROPDOWN,
  value,
});

export type SearchCompanyByNameAction = ReturnType<typeof searchCompanyByNameAction>;
export const searchCompanyByNameAction = (value: string) => ({
  type: SEARCH_COMPANY_BY_NAME as typeof SEARCH_COMPANY_BY_NAME,
  value,
});

export type AgentsModuleAction =
  | GetTravelAgentRequestAction
  | GetTravelAgentSuccessAction
  | GetTravelAgentFailureAction
  | SelectedTaChangeAction
  | ShowTaDropdownAction
  | SearchTaByNameChangeAction
  | GetCompaniesRequestAction
  | GetCompaniesSuccessAction
  | GetCompaniesFailureAction
  | SelectedCompanyChangeAction
  | ShowCompanyDropdownAction
  | SearchCompanyByNameAction;
