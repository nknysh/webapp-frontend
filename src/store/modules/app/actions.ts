export const SET_LATEST_APP_VERSION = 'app/SET_LATEST_APP_VERSION';

export type SetLatestAppVersion = ReturnType<typeof setLatestAppVersion>;
export const setLatestAppVersion = (latestAppVersion: string, isAppDeprecated: boolean) => ({
  type: SET_LATEST_APP_VERSION as typeof SET_LATEST_APP_VERSION,
  latestAppVersion,
  isAppDeprecated
});


export type AppAction =
 | SetLatestAppVersion
