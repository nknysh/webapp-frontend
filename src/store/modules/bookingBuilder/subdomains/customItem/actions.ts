export const SHOW_CUSTOM_ITEM_FORM = 'bookingBuilder/customItem/SHOW_CUSTOM_ITEM_FORM';
export const HIDE_CUSTOM_ITEM_FORM = 'bookingBuilder/customItem/HIDE_CUSTOM_ITEM_FORM';

export const UPDATE_CUSTOM_ITEM_NAME = 'bookingBuilder/customItem/UPDATE_CUSTOM_ITEM_NAME';
export const UPDATE_CUSTOM_ITEM_TOTAL = 'bookingBuilder/customItem/UPDATE_CUSTOM_ITEM_TOTAL';
export const UPDATE_CUSTOM_ITEM_DESCRIPTION = 'bookingBuilder/customItem/UPDATE_CUSTOM_ITEM_DESCRIPTION';
export const UPDATE_CUSTOM_ITEM_COUNTS_AS_MEAL_PLAN = 'bookingBuilder/customItem/UPDATE_CUSTOM_ITEM_COUNTS_AS_MEAL_PLAN';
export const UPDATE_CUSTOM_ITEM_COUNTS_AS_TRANSFER = 'bookingBuilder/customItem/UPDATE_CUSTOM_ITEM_COUNTS_AS_MEAL_PLAN';


export type ShowCustomItemFormAction = ReturnType<typeof showCustomItemFormAction>;
export const showCustomItemFormAction = () => ({
  type: SHOW_CUSTOM_ITEM_FORM as typeof SHOW_CUSTOM_ITEM_FORM
});

export type HideCustomItemFormAction = ReturnType<typeof hideCustomItemFormAction>;
export const hideCustomItemFormAction = () => ({
  type: HIDE_CUSTOM_ITEM_FORM as typeof HIDE_CUSTOM_ITEM_FORM
});

//------------------- edit ----------------------------------

export type UpdateCustomItemNameAction = ReturnType<typeof updateCustomItemNameAction>;
export const updateCustomItemNameAction = (value: string) => ({
  type: UPDATE_CUSTOM_ITEM_NAME as typeof UPDATE_CUSTOM_ITEM_NAME,
  value
});

export type UpdateCustomItemTotalAction = ReturnType<typeof updateCustomItemTotalAction>;
export const updateCustomItemTotalAction = (value: string) => ({
  type: UPDATE_CUSTOM_ITEM_TOTAL as typeof UPDATE_CUSTOM_ITEM_TOTAL,
  value
});

export type UpdateCustomItemDescriptionAction = ReturnType<typeof updateCustomItemDescriptionAction>;
export const updateCustomItemDescriptionAction = (value: string) => ({
  type: UPDATE_CUSTOM_ITEM_DESCRIPTION as typeof UPDATE_CUSTOM_ITEM_DESCRIPTION,
  value
});

export type UpdateCustomItemCountsAsMealPlanAction = ReturnType<typeof updateCustomItemCountsAsMealPlanAction>;
export const updateCustomItemCountsAsMealPlanAction = (value: boolean) => ({
  type: UPDATE_CUSTOM_ITEM_COUNTS_AS_MEAL_PLAN as typeof UPDATE_CUSTOM_ITEM_COUNTS_AS_MEAL_PLAN,
  value
});

export type UpdateCustomItemCountsAsTransferAction = ReturnType<typeof updateCustomItemCountsAsTransferAction>;
export const updateCustomItemCountsAsTransferAction = (value: boolean) => ({
  type: UPDATE_CUSTOM_ITEM_COUNTS_AS_TRANSFER as typeof UPDATE_CUSTOM_ITEM_COUNTS_AS_TRANSFER,
  value
});


export type CustomItemAction =
  | ShowCustomItemFormAction
  | HideCustomItemFormAction

  | UpdateCustomItemNameAction
  | UpdateCustomItemTotalAction
  | UpdateCustomItemDescriptionAction
  | UpdateCustomItemCountsAsMealPlanAction
  | UpdateCustomItemCountsAsTransferAction;
