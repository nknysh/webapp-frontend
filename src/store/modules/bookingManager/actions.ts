import { IBooking } from 'services/BackendApi';

export const GET_BOOKING_REQUEST = 'bookingManager/GET_BOOKING_REQUEST';
export const GET_BOOKING_SUCCESS = 'bookingManager/GET_BOOKING_SUCCESS';
export const GET_BOOKING_FAILURE = 'bookingManager/GET_BOOKING_FAILURE';

export const REQUEST_TO_BOOK_REQUEST = 'bookingManager/REQUEST_TO_BOOK_REQUEST';
export const REQUEST_TO_BOOK_SUCCESS = 'bookingManager/REQUEST_TO_BOOK_SUCCESS';
export const REQUEST_TO_BOOK_FAILURE = 'bookingManager/REQUEST_TO_BOOK_FAILURE';

export const CANCEL_REQUEST = 'bookingManager/CANCEL_REQUEST';
export const CANCEL_SUCCESS = 'bookingManager/CANCEL_SUCCESS';
export const CANCEL_FAILURE = 'bookingManager/CANCEL_FAILURE';

export const CONFIRM_REQUEST = 'bookingManager/CONFIRM_REQUEST';
export const CONFIRM_SUCCESS = 'bookingManager/CONFIRM_SUCCESS';
export const CONFIRM_FAILURE = 'bookingManager/CONFIRM_FAILURE';

export type GetBookingRequestAction = ReturnType<typeof getBookingRequestAction>;
export const getBookingRequestAction = (uuid: string) => ({
  type: GET_BOOKING_REQUEST as typeof GET_BOOKING_REQUEST,
  uuid,
});

export type GetBookingSuccessAction = ReturnType<typeof getBookingSuccessAction>;
export const getBookingSuccessAction = (booking: IBooking) => ({
  type: GET_BOOKING_SUCCESS as typeof GET_BOOKING_SUCCESS,
  booking,
});

export type GetBookingFailureAction = ReturnType<typeof getBookingFailureAction>;
export const getBookingFailureAction = (error: string) => ({
  type: GET_BOOKING_FAILURE as typeof GET_BOOKING_FAILURE,
  error,
});

// Request to Book actions
export type RequestToBookRequestAction = ReturnType<typeof requestToBookRequestAction>;
export const requestToBookRequestAction = (booking: IBooking) => ({
  type: REQUEST_TO_BOOK_REQUEST as typeof REQUEST_TO_BOOK_REQUEST,
  booking,
});
export type RequestToBookSuccessAction = ReturnType<typeof requestToBookSuccessAction>;
export const requestToBookSuccessAction = (booking: IBooking) => ({
  type: REQUEST_TO_BOOK_SUCCESS as typeof REQUEST_TO_BOOK_SUCCESS,
  booking,
});
export type RequestToBookFailureAction = ReturnType<typeof requestToBookFailureAction>;
export const requestToBookFailureAction = error => ({
  type: REQUEST_TO_BOOK_FAILURE as typeof REQUEST_TO_BOOK_FAILURE,
  error,
});

// Cancel actions
export type CancelRequestAction = ReturnType<typeof cancelRequestAction>;
export const cancelRequestAction = (booking: IBooking) => ({
  type: CANCEL_REQUEST as typeof CANCEL_REQUEST,
  booking,
});
export type CancelSuccessAction = ReturnType<typeof cancelSuccessAction>;
export const cancelSuccessAction = (booking: IBooking) => ({
  type: CANCEL_SUCCESS as typeof CANCEL_SUCCESS,
  booking,
});
export type CancelFailureAction = ReturnType<typeof cancelFailureAction>;
export const cancelFailureAction = error => ({
  type: CANCEL_FAILURE as typeof CANCEL_FAILURE,
  error,
});

// confirm actions
export type ConfirmRequestAction = ReturnType<typeof confirmRequestAction>;
export const confirmRequestAction = (uuid: string) => ({
  type: CONFIRM_REQUEST as typeof CONFIRM_REQUEST,
  uuid,
});
export type ConfirmSuccessAction = ReturnType<typeof confirmSuccessAction>;
export const confirmSuccessAction = (booking: IBooking) => ({
  type: CONFIRM_SUCCESS as typeof CONFIRM_SUCCESS,
  booking,
});
export type ConfirmFailureAction = ReturnType<typeof confirmFailureAction>;
export const confirmFailureAction = error => ({
  type: CONFIRM_FAILURE as typeof CONFIRM_FAILURE,
  error,
});

export type BookingManagerAction =
  | GetBookingRequestAction
  | GetBookingSuccessAction
  | GetBookingFailureAction
  | RequestToBookRequestAction
  | RequestToBookSuccessAction
  | RequestToBookFailureAction
  | CancelRequestAction
  | CancelSuccessAction
  | CancelFailureAction
  | ConfirmRequestAction
  | ConfirmSuccessAction
  | ConfirmFailureAction;
