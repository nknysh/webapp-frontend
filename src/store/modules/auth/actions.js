import { prop, propEq, omit, propOr, mergeDeepLeft, equals } from 'ramda';
import { FORBIDDEN } from 'http-status';

import client from 'api/auth';
import userClient from 'api/users';
import { UserStatusTypes, AuthTypes } from 'config/enums';

import { genericAction, successAction, errorAction, errorFromResponse, storeReset } from 'store/common';
import { enqueueNotification } from 'store/modules/ui/actions';
import { getCurrentUserType } from './selectors';

export const AUTH_REQUEST = 'AUTH_REQUEST';
export const AUTH_CHECK = 'AUTH_CHECK';
export const AUTH_OK = 'AUTH_OK';
export const AUTH_RESET = 'AUTH_RESET';
export const AUTH_SET_TOKEN = 'AUTH_SET_TOKEN';
export const AUTH_SIGN_UP = 'AUTH_SIGN_UP';
export const AUTH_LOG_OUT = 'AUTH_LOG_OUT';
export const AUTH_PASSWORD_RESET = 'AUTH_PASSWORD_RESET';
export const AUTH_SET_PASSWORD = 'AUTH_SET_PASSWORD';
export const AUTH_COUNTRY_SET = 'AUTH_COUNTRY_SET';
export const AUTH_PASSWORD_UPDATE = 'AUTH_PASSWORD_UPDATE';

// Localstorage constants for auth
export const AUTH_TOKEN = 'authToken';
export const AUTH_USER = 'authUser';
export const AUTH_COUNTRY_CODE = 'authCountryCode';

/**
 * Auth reset action
 *
 * @returns {object}
 */
const authReset = () => ({
  type: AUTH_RESET,
});

/**
 * Set token action
 *
 * @param {string} token
 * @returns {object}
 */
export const setToken = token => ({
  type: AUTH_SET_TOKEN,
  payload: { token },
});

/**
 * Auth request action
 *
 * @param {*} value
 * @returns {object}
 */
export const authRequest = value => ({
  type: AUTH_REQUEST,
  payload: value,
});

/**
 * Auth password reset action
 *
 * @param {*} value
 * @returns {object}
 */
export const authPasswordReset = value => ({
  type: AUTH_PASSWORD_RESET,
  payload: value,
});

/**
 * Auth set password reset action
 *
 * @param {*} value
 * @returns {object}
 */
export const authSetPasswordReset = value => ({
  type: AUTH_SET_PASSWORD,
  payload: value,
});

/**
 * Auth ok action
 *
 * @param {*} values
 * @returns {object}
 */
export const authOk = values => ({
  type: AUTH_OK,
  payload: values,
});

/**
 * Auth sign up action
 *
 * @param {*} values
 * @returns {object}
 */
export const authSignUp = values => ({
  type: AUTH_SIGN_UP,
  payload: values,
});

/**
 * Auth log out action
 *
 * @param {string} token
 * @returns {object}
 */
export const authLogOut = token => ({
  type: AUTH_LOG_OUT,
  payload: token,
});

/**
 * Auth set country action
 *
 * @param {string} countryCode
 * @returns {object}
 */
export const authSetCountryAction = countryCode => ({
  type: AUTH_COUNTRY_SET,
  payload: countryCode,
});

/**
 * Set remembered token
 *
 * @param {string} token
 */
export const setRememberedToken = token => localStorage.setItem(AUTH_TOKEN, token);

/**
 * Set remembered user
 *
 * @param {object} user
 */
export const setRememberedUser = user => localStorage.setItem(AUTH_USER, JSON.stringify(user));

/**
 * Set remembered country
 *
 * @param {string} countryCode
 */
export const setRememberedCountry = countryCode => localStorage.setItem(AUTH_COUNTRY_CODE, countryCode);

/**
 * Delete remembered token
 */
export const deleteRememberedToken = () => localStorage.removeItem(AUTH_TOKEN);

/**
 * Delete remembered user
 */
export const deleteRememberedUser = () => localStorage.removeItem(AUTH_USER);

/**
 * Delete remembered country
 */
export const deleteRememberedCountry = () => localStorage.removeItem(AUTH_COUNTRY_CODE);

/**
 * Auth set country action
 *
 * Sets the remembered country into local storage
 *
 * @param {string} countryCode
 * @returns {Function}
 */
export const authSetCountry = countryCode => dispatch => {
  setRememberedCountry(countryCode);
  dispatch(authSetCountryAction(countryCode));
};

/**
 * Get user action
 *
 * @param {Function} dispatch
 */
export const getUser = async dispatch => {
  dispatch(authRequest());

  try {
    const { data } = await client.getUser();
    dispatch(successAction(AUTH_REQUEST, data));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_REQUEST, e, 'There was a problem fetching the current user.'));
  }
};

/**
 * Sign up action
 *
 * @param {object} values
 * @returns {Function}
 */
export const signUp = values => async dispatch => {
  dispatch(authSignUp(values));

  try {
    const { data } = await client.signUp({ data: { attributes: values } });
    dispatch(successAction(AUTH_SIGN_UP, data));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_SIGN_UP, e));
  }
};

/**
 * Clear user action
 *
 * Removes user form local storage
 *
 * @param {Function} dispatch
 */
const clearUser = dispatch => {
  deleteRememberedToken();
  deleteRememberedUser();
  deleteRememberedCountry();

  dispatch(authReset());
  dispatch(successAction(AUTH_LOG_OUT, {}));
  dispatch(storeReset());
};

/**
 * Log out action
 *
 * Trrigger logout request
 *
 * @param {string} token
 * @returns {Function}
 */
export const logOut = token => async dispatch => {
  dispatch(authLogOut(token));

  clearUser(dispatch);

  try {
    await client.logOut();
  } catch (e) {
    return true;
  }
};

/**
 * Persist user action
 *
 * Sets user data to local storage
 *
 * @param {Function} dispatch
 * @param {object} data
 */
const persistUser = (dispatch, data) => {
  const userUuid = prop('uuid', data);

  // If a user is 'pending' then throw a forbidden error and return
  if (propEq('status', UserStatusTypes.PENDING, data)) {
    return dispatch(errorAction(AUTH_REQUEST, { status: FORBIDDEN, unverified: true }));
  }

  setRememberedToken(userUuid);
  setRememberedUser(data);
  setRememberedCountry(propOr('GB', 'countryCode', data));

  dispatch(setToken(userUuid));
};

/**
 * Log in action
 *
 * @param {object} values
 * @returns {Function}
 */
export const logIn = values => async dispatch => {
  dispatch(authRequest(omit(['password'], values)));

  try {
    const {
      data: { data },
    } = await client.logIn({ data: { attributes: values } });

    persistUser(dispatch, data);
    dispatch(successAction(AUTH_REQUEST, { user: { ...data } }));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_REQUEST, e));
  }
};

/**
 * Reset password action
 *
 * @param {object} values
 * @returns {Function}
 */
export const resetPassword = values => async dispatch => {
  dispatch(authPasswordReset(values));

  try {
    await client.resetPassword({ data: { attributes: values } });
    dispatch(successAction(AUTH_PASSWORD_RESET, values));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_SET_PASSWORD, e));
    throw e;
  }
};

/**
 * Set password action
 *
 * @param {object} values
 * @returns {Function}
 */
export const setPassword = values => async dispatch => {
  dispatch(authSetPasswordReset(omit(['values'], values)));

  try {
    await client.setPassword({ data: { attributes: values } });
    dispatch(successAction(AUTH_SET_PASSWORD, values));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_SET_PASSWORD, e));
  }
};

/**
 * Auth check action
 *
 * @param {object} params
 * @returns {Function}
 */
export const authCheck = params => async (dispatch, getState) => {
  dispatch(genericAction(AUTH_CHECK));

  const role = getCurrentUserType(getState());

  // Associations for auth call are different per role
  // TAs : assignedSalesRepresentatives
  // Everyone else : assignedTravelAgents
  const authParams = mergeDeepLeft(params, {
    associations: equals(AuthTypes.TA, role) ? 'assignedSalesRepresentatives' : 'assignedTravelAgents',
  });

  try {
    const {
      data: { data },
    } = await userClient.me(authParams);
    dispatch(successAction(AUTH_CHECK, { user: { ...data } }));
    persistUser(dispatch, data);
  } catch (e) {
    clearUser(dispatch);

    throw e;
  }
};

/**
 * Update password action
 *
 * @param {object} body
 * @returns {Function}
 */
export const updatePassword = body => async dispatch => {
  dispatch(genericAction(AUTH_PASSWORD_UPDATE, body));

  try {
    await client.updatePassword({ data: { attributes: body } });
    dispatch(successAction(AUTH_SET_PASSWORD, body));
    dispatch(enqueueNotification({ message: 'Password has been updated.', options: { variant: 'success' } }));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_PASSWORD_UPDATE, e));
  }
};
