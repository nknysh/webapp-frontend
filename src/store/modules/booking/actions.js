import { __, pick, lensProp, lensPath, prop, set, view, map, mapObjIndexed, pipe, mergeDeepRight } from 'ramda';
import { eachDay, subDays } from 'date-fns';

import client from 'api/bookings';
import { formatDate } from 'utils';

import { successAction, errorFromResponse } from 'store/common';
import { enqueueNotification } from 'store/modules/ui/actions';
import { getHotelRoom } from 'store/modules/hotels/selectors';

import { getBookingData, getBookingByHotelId, getBookingRoomDatesById } from './selectors';

const accommodationProductsLens = lensProp('accommodationProducts');
const datesLens = lensProp('dates');
const ratesLens = lensPath(['dates', 'rates']);

export const BOOKING_UPDATE = 'BOOKING_UPDATE';
export const BOOKING_SUBMIT = 'BOOKING_SUBMIT';
export const BOOKING_REMOVE = 'BOOKING_REMOVE';

export const updateBookingAction = payload => ({
  type: BOOKING_UPDATE,
  payload,
});

export const submitBookingAction = payload => ({
  type: BOOKING_SUBMIT,
  payload,
});

export const removeBooking = payload => ({
  type: BOOKING_REMOVE,
  payload,
});

export const updateBooking = (id, payload) => (dispatch, getState) => {
  const state = getState();

  dispatch(updateBookingAction(payload));
  dispatch(successAction(BOOKING_UPDATE, mergeDeepRight(getBookingData(state), { [id]: payload })));
};

export const completeBooking = (id, payload) => async (dispatch, getState) => {
  dispatch(updateBooking(id, payload));

  const state = getState();
  const booking = getBookingByHotelId(state, id);

  const finaliseAccommodationProducts = (product, productId) => {
    const hotelRoom = getHotelRoom(state, id, productId);
    const { to, from } = getBookingRoomDatesById(state, id, productId);

    const rates = prop('rates', hotelRoom);

    const days = eachDay(from, subDays(to, 1));
    const ratesForDays = pipe(
      map(formatDate),
      pick(__, rates)
    )(days);

    return pipe(
      set(datesLens, { to, from }),
      set(ratesLens, ratesForDays)
    )(product);
  };

  const finalBooking = pipe(
    set(
      accommodationProductsLens,
      mapObjIndexed(finaliseAccommodationProducts, view(accommodationProductsLens, booking))
    )
  )(booking);

  dispatch(submitBookingAction(finalBooking));

  try {
    const {
      data: { data },
    } = await client.createBooking({ data: { attributes: { id, ...finalBooking } } });

    dispatch(updateBooking(id, { reservationId: prop('uuid', data) }));
    dispatch(successAction(BOOKING_SUBMIT, data));
  } catch (e) {
    dispatch(
      enqueueNotification({ message: 'There was a problem creating your booking.', options: { variant: 'error' } })
    );
    dispatch(errorFromResponse(BOOKING_SUBMIT, e));
    throw e;
  }
};
