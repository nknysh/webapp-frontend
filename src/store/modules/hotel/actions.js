import { pathOr } from 'ramda';

import { formatDate } from 'utils';

import client from 'api/hotels';

import { successAction, errorFromResponse } from 'store/common';
import { setHotels } from 'store/modules/hotels/actions';
import { getSearchQuery } from 'store/modules/search/selectors';
import { enqueueNotification } from 'store/modules/ui/actions';

export const HOTEL = 'HOTEL';
export const HOTEL_ROOMS = 'HOTEL_ROOMS';

export const fetchHotelAction = () => ({
  type: HOTEL,
});

export const fetchHotelRoomsAction = payload => ({
  type: HOTEL_ROOMS,
  payload,
});

export const fetchHotel = id => async (dispatch, getState) => {
  const state = getState();
  const searchQuery = getSearchQuery(state);

  const startDate = pathOr(Date.now(), ['dates', 'from'], searchQuery);
  const endDate = pathOr(Date.now(), ['dates', 'to'], searchQuery);

  dispatch(fetchHotelAction());

  try {
    const {
      data: { data },
    } = await client.getHotel(id, {
      startDate: formatDate(startDate),
      endDate: formatDate(endDate),
    });

    dispatch(successAction(HOTEL, data));
    dispatch(setHotels(data));
  } catch (e) {
    dispatch(errorFromResponse(HOTEL, e));
    dispatch(
      enqueueNotification({
        message: 'There was a problem fetching this hotel. Please try again.',
        options: { variant: 'error' },
      })
    );
  }
};
