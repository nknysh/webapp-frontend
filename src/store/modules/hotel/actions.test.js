import { FETCH_HOTEL, fetchHotelAction, fetchHotel } from './actions';

describe('pages actions', () => {
  describe('fetchHotelAction', () => {
    it('returns action', () => {
      expect(fetchHotelAction('foo')).toEqual({
        type: FETCH_HOTEL,
      });
    });
  });
  describe('fetchHotel', () => {
    it('calls dispatch thunks', () => {
      const dispatch = jest.fn();

      fetchHotel()(dispatch, () => {});

      expect(dispatch).toHaveBeenCalledTimes(1);
    });
  });
});
