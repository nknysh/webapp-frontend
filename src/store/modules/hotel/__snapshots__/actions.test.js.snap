// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hotel actions fetchHotelAction returns action 1`] = `
Object {
  "type": "HOTEL",
}
`;
